### PR TITLE
Implementation of Spatial Pyramid Pooling

### PR DESCRIPTION
**Pull request by Susmit-A**

### CODE REVIEW

File: tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for AdaptivePooling layers."""
+
+import pytest
+from tensorflow_addons.layers.spatial_pyramid_pooling import SpatialPyramidPooling2D
+
```
Comment (ashutosh1919) [inline tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py:19]: Please consider adding more tests around config, serializability, dtype and functionality checks.

Comment (Susmit-A) [reply tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py:24]: I've added the tests.

File: .github/CODEOWNERS [hunk 2]
```diff
@@ -71,6 +71,8 @@
 /tensorflow_addons/image/translate_ops.py @sayoojbk
 /tensorflow_addons/image/tests/translate_ops_test.py @sayoojbk
 
+/tensorflow_addons/layers/spatial_pyramid_pooling.py @Susmit-A
```
Comment (AakashKumarNain) [inline .github/CODEOWNERS:74]: Order alphabetically. Move this below `sparsemax`

File: tensorflow_addons/layers/spatial_pyramid_pooling.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+from typeguard import typechecked
+from typing import Union, Iterable
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class SpatialPyramidPooling2D(tf.keras.layers.Layer):
+    """Performs Spatial Pyramid Pooling.
+
+    Original Paper: https://arxiv.org/pdf/1406.4729.pdf
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/spatial_pyramid_pooling.py:29]: A little more description, like when and why is it useful?

**Comment by Susmit-A** [comment]
The tests show one cancelled and one skipped check. I can't find any logs regarding the same. Is it an issue with my code?

File: tensorflow_addons/layers/spatial_pyramid_pooling.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+    Spatial Pyramid Pooling generates a fixed-length representation
+    regardless of input size/scale. It is typically used before layers
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/spatial_pyramid_pooling.py:32]: It is typically used before a layer that expects a fixed input shape, e.g. a `Dense` layer.

File: tensorflow_addons/layers/spatial_pyramid_pooling.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    that require a constant input shape, for example before Dense Layers.
+
+    Arguments:
+      bins: Either a collection of integers or a collection of collections of 2 integers.
+        Each element in the inner collection must contain 2 integers, (pooled_rows, pooled_cols)
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/spatial_pyramid_pooling.py:37]: Put an example of collection like `[1, 2, 3]` or  `[(1, 1), (2, 2), (3, 3)]`

File: tensorflow_addons/layers/spatial_pyramid_pooling.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            self.pool_layers.append(AdaptiveAveragePooling2D(bin, self.data_format))
+        super().__init__(*args, **kwargs)
+
+    def call(self, inputs, **kwargs):
+        dynamic_input_shape = tf.shape(inputs)
+        outputs = []
+        index = 0
+        if self.data_format == "channels_last":
+            for bin in self.bins:
+                new_inp = inputs[
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/spatial_pyramid_pooling.py:80]: The slices are not well readable. Can you refactor this to make it please?

File: tensorflow_addons/layers/spatial_pyramid_pooling.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                output = self.pool_layers[index](new_inp)
+                output = tf.reshape(
+                    output, [dynamic_input_shape[0], bin[0] * bin[1], inputs.shape[-1]]
+                )
+                outputs.append(output)
+                index += 1
+            outputs = tf.concat(outputs, axis=1)
+        else:
+            for bin in self.bins:
+                new_inp = inputs[
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/spatial_pyramid_pooling.py:95]: Same here for readability

File: tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+def test_keras():
+    test_inputs = np.arange(start=0.0, stop=16.0, step=1.0).astype(np.float32)
+    test_inputs = np.reshape(test_inputs, (1, 4, 4, 1))
+    test_output = [[[7.5], [2.5], [4.5], [10.5], [12.5]]]
+
+    inputs = tf.keras.layers.Input((None, None, 1))
+    spp = SpatialPyramidPooling2D([1, 2])(inputs)
+    model = tf.keras.Model(inputs=[inputs], outputs=[spp])
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        model.save(tmp_dir + "/spp_model.h5")
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py:78]: Use `pathlib` to define paths

File: tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        model = tf.keras.models.load_model(tmp_dir + "/spp_model.h5")
```
Comment (AakashKumarNain) [inline tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py:79]: Same here for path

File: tensorflow_addons/layers/spatial_pyramid_pooling.py (hunk 6 above)
Comment (Susmit-A) [reply tensorflow_addons/layers/spatial_pyramid_pooling.py:80]: Do you mean using `tf.slice` or simply expanding the internal terms? TensorFlow documentation recommends using the pythonic way.

Comment (AakashKumarNain) [reply tensorflow_addons/layers/spatial_pyramid_pooling.py:89]: No, I mean the way the code is laid out, it's less readable.

**Comment by AakashKumarNain** [comment]
@gabrieldemarmiesse @seanpmorgan can you please take a look at this one whenever you are free? Thanks

**Comment by Susmit-A** [comment]
@gabrieldemarmiesse @seanpmorgan @facaiy @ashutosh1919 Please review the PR.

File: tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+@pytest.mark.usefixtures("maybe_run_functions_eagerly")
+def test_keras():
+    test_inputs = np.arange(start=0.0, stop=16.0, step=1.0).astype(np.float32)
+    test_inputs = np.reshape(test_inputs, (1, 4, 4, 1))
+    test_output = [[[7.5], [2.5], [4.5], [10.5], [12.5]]]
+
+    inputs = tf.keras.layers.Input((None, None, 1))
+    spp = SpatialPyramidPooling2D([1, 2])(inputs)
+    model = tf.keras.Model(inputs=[inputs], outputs=[spp])
+    with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py:78]: Please use the pytest `tmp_dir` to handle this:
https://docs.pytest.org/en/latest/tmpdir.html

Example:
https://github.com/tensorflow/addons/commit/b48051a662e1dad22f745e3d81cebc477e5b2906#diff-1a636963525121cf08972a4424360cffR34

Comment (Susmit-A) [reply tensorflow_addons/layers/tests/spatial_pyramid_pooling_test.py]: I've made the requested changes.

**Pull request merged**